Services.Test/CustomDeviceModelsTest.cs

```diff
@@ -0,0 +1,187 @@
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.StorageAdapter;
+using Moq;
+using Newtonsoft.Json;
+using Services.Test.helpers;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Services.Test
+{
+    public class CustomDeviceModelsTest
+    {
+        private const string STORAGE_COLLECTION = "deviceModels";
+
+        private readonly ITestOutputHelper log;
+        private readonly Mock<IStorageAdapterClient> storage;
+        private readonly Mock<ILogger> logger;
+        private readonly CustomDeviceModels target;
+
+        public CustomDeviceModelsTest(ITestOutputHelper log)
+        {
+            this.log = log;
+
+            this.storage = new Mock<IStorageAdapterClient>();
+            this.logger = new Mock<ILogger>();
+
+            this.target = new CustomDeviceModels(
+                this.storage.Object,
+                this.logger.Object);
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void InitialListIsEmpty()
+        {
+            // Arrange
+            this.ThereAreNoCustomDeviceModelsInStorage();
+
+            // Act
+            var result = this.target.GetListAsync().Result;
+
+            // Assert
+            Assert.Empty(result);
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CreateCustomDeviceModel()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { ETag = "Etag_1" };
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            DeviceModel result = this.target.InsertAsync(deviceModel).Result;
+
+            // Assert
+            Assert.NotNull(result);
+            Assert.Equal(deviceModel.ETag, result.ETag);
+        }
+
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CreatedCustomDeviceModelsAreStored()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { ETag = "Etag_1" };
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            this.target.InsertAsync(deviceModel).Wait();
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                null));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CustomDeviceModelsCanBeUpserted()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { Id= "id", ETag = "oldEtag" };
+            var updatedDeviceModel = new DeviceModel  { ETag = "newETag" };
+
+            this.SaveDeviceModelInStorage(deviceModel);
+            this.UpdateDeviceModelInStorage(updatedDeviceModel);
+
+            // Act
+            this.target.UpsertAsync(deviceModel).Wait();
```

Comment:
when using .Wait() in tests, remember to set a timeout, e.g. Timespan.FromSeconds(10), to avoid CI builds hanging

Reply:
Done.